rust/kernel/lib.rs

```diff
@@ -21,6 +21,7 @@
     const_panic,
     const_raw_ptr_deref,
     const_unreachable_unchecked,
+    receiver_trait,
```

Comment:
This feature is not meant to eventually be stabilized. It doesn't have a tracking issue and `Receiver` is marked as `#[doc(hidden)]`.

Reply:
Given that we are going to have a in-tree copy of `alloc` crate I think it's okay to use it for now. We'll need `DispatchFromDyn` anyway which is also hidden. I think eventually we should push for a stable way in Rust to define custom smart pointers that are as usable as `Arc`.

Reply:
`Receiver` could be stabilized if wanted. For `DispatchFromDyn` this is not as easy as by definition it relies on an unstable implementation detail. An incorrect implementation of it can cause an ICE or miscompilation I think. Basically for `Foo` to have a valid `DispatchFromDyn` impl, `Foo<dyn Trait>` must consist of only a single fat pointer with the data part being identical to the thin pointer in `Foo<ConcreteType>`. Rustc assumes that `Foo<dyn Trait>` is passed in two registers. The second register contains the vtable and is used to get the method function pointer. The first register is then passed as sole argument to the method. Turning `Foo<ConcreteType>` into `Foo<dyn Trait>` also depends on another unstable feature called `CoerceUnsized`, which itself depends on `Unsize`.

Reply:
To add some more context to what Gary already said: we had a technical meeting on the `alloc` crate last week. The consensus was that we'll have a "fork" of `alloc` in the kernel tree; changes directly to it are discouraged but in the event that they are needed, we will work with the Rust community to upstream them so in steady-state our fork and upstream should be the same. 

A replacement for `Arc` was explicitly discussed, and @joshtriplett mentioned that it may be reasonable for Rust to allow `Arc`-like implementations that are not so tied to the compiler, and that we should explore this as a way to minimise changes to our fork of `alloc`. While this isn't a reality, and once our fork is committed, we can move this `Ref` implementation to our fork of `alloc` if we want to minimise the risk of other components accidentally relying on unstable features required by `Ref`.